### DISCUSSION TITLE
Some multi-line docstrings are not stripping leading block whitespace

### DISCUSSION THREAD

**Discussion by mzealey** [fastapi/fastapi#12174] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
fastapi_app = FastAPI(                                         
        description="""                                                                        
            abc.
              
            cde.
            """
    )
```

### Description

Open /redoc and it displays this as code (ie as if it were being indented by the 12 spaces).

It seems that when this style description is added to functions as docstring, or some other places, the block-indent stripping occurs correctly. However when used in the `FastAPI(description=)` field, or as `openapi_tags` descriptions the block-indent stripping does not occur.

Interestingly the annotation around the FastAPI `description` field shows a use-case exactly like this which I assume does not work correctly.

### Operating System

Linux

### Operating System Details

in docker container

### FastAPI Version

0.114.0

### Pydantic Version

2.9.1

### Python Version

3.12.5

**Comment by denisw** [comment, marked as answer]
The most straightforward workaround is to manually apply [`inspect.cleandoc()`](https://docs.python.org/3/library/inspect.html#inspect.cleandoc) to the description before passing it to FastAPI. This is in fact the very function that FastAPI applies to route function docstrings automatically.

```python
from inpect import cleandoc


fastapi_app = FastAPI(                                         
        description=cleandoc(
            """                                                                        
            abc.
              
            cde.
            """
        )
    )
```

**Comment by yakubka** [comment]
FastAPI uses the function docstring as the operation description in the OpenAPI spec. It strips leading and trailing whitespace from the entire docstring but does not normalize the indentation of individual lines the way `inspect.cleandoc` does.

If your docstring has a consistent indent (which is the standard Python style), the raw string will have leading spaces on every line except the first, and those spaces appear in the rendered Swagger UI or ReDoc description.

The fix is to pass the docstring through `inspect.cleandoc` before it reaches the spec, or to write the docstring with no indent after the first line.

```python
import inspect

def my_route():
    """
    First line.

    Second paragraph with details.
    """
    ...

# FastAPI reads __doc__ directly; you can post-process it:
my_route.__doc__ = inspect.cleandoc(my_route.__doc__ or "")
```

Alternatively, open a PR to FastAPI to apply `inspect.cleandoc` to `func.__doc__` inside `get_openapi_operation_metadata`, which is where the description is extracted.

**Comment by alwaysprince05** [comment]
Hi @yakubka, I'd like to pick this up if no one else is on it.

I checked, and the spot you mentioned is already handled — route docstrings
get inspect.cleandoc at fastapi/routing.py:1062. The actual gap is at the app
level:
- fastapi/applications.py:876  (self.description)
- fastapi/applications.py:882  (self.openapi_tags)

I reproduced it on master (0.139.2): indented FastAPI(description=...) shows
up in the spec with leading spaces intact.

Plan: apply inspect.cleandoc to the app description and to each
openapi_tags description at construction, skipping None/empty. Plus a couple
of regression tests.

Two quick questions:
- Also clean summary/terms_of_service, or just description + tags?
- Clean at construction (matches routing.py:1062) or lazily at app.openapi()?

Won't open a PR until you give the go-ahead.

**Comment by dannymaaz** [comment]
I reproduced this behavior on the current `master` branch.

Application descriptions and tag descriptions are currently passed to the generated OpenAPI schema without removing indentation from multiline strings. This can cause Markdown content written with triple-quoted strings to be rendered as code blocks.

A focused solution could:

* apply `inspect.cleandoc()` to the application description before adding it to the OpenAPI schema;
* apply the same normalization to tag descriptions;
* avoid mutating the original `openapi_tags` objects provided by the user;
* add regression tests for multiline application and tag descriptions;
* preserve relative indentation inside nested Markdown content, such as lists and code blocks.

Single-line descriptions and already normalized strings would remain unchanged.

Would this approach be acceptable for a pull request? I would be happy to implement it with regression coverage.

**Comment by dannymaaz** [comment]
I reproduced this behavior on the current `master` branch and prepared a focused implementation with regression coverage.

The proposed change:

- cleans multiline application descriptions before adding them to the OpenAPI schema;
- applies the same behavior to tag descriptions;
- preserves single-line descriptions unchanged;
- preserves relative indentation inside Markdown lists and fenced code blocks;
- avoids mutating the original `openapi_tags` objects.

The implementation uses `inspect.cleandoc()`, matching the behavior already used for path operation docstrings.

The implementation is ready in my fork. Would this approach be acceptable for a pull request?
